### PR TITLE
Make system fonts work more reliably

### PR DESCRIPTION
**Pull request by r00ster91**
So for a while I tried to use the Commodore 64 font (the same that VVVVVV uses) in my favorite messenger.
But I always had this problem that I had the font installed and I entered the exact name of the font but it didn't work:

image
<div align="center"><i>Why doesn't it work :(</i></div>

<br>

Well it turns out that the problem was the space in the font's name:

image
<div align="center"><i>It works!</i></div>

<br>

Doing this automatically for the user should avoid a lot of confusion.

```
Signed-off-by: r00ster91 <r00ster91@proton.me>
```

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Comment by t3chguy** [comment]
This will need to be a bit smarter as you can specify multiple comma separated fonts. E.g `Inter, Twemoji`

**Comment by r00ster91** [comment]
@t3chguy I wonder how well this would work in practice?
```ts
`"${"Fira Sans Thin, Commodore 64".split(',').map(font => font.trim()).join("\",\"")}"` // "Fira Sans Thin","Commodore 64"
```
Alternatively, we could only do ``"${font}"`` if there is no comma included in the font string (i.e. only one font)

**Comment by t3chguy** [comment]
The split map join is what I had in mind but please add the double quotes in the map phase to aid readability

**Comment by t3chguy** [comment]
The split map join is what I had in mind but please add the double quotes in the map phase to aid readability

**Comment by t3chguy** [comment]
Also worth considering that existing users may have already added double quotes manually

**Comment by r00ster91** [comment]
I wonder how it is now?

File: src/settings/watchers/FontWatcher.ts [hunk 1]
```diff
@@ -64,6 +64,17 @@ export class FontWatcher implements IWatcher {
...
+            document.body.style.fontFamily = font
+                .split(',')
+                .map(font => {
+                    font = font.trim();
+                    if (!font.startsWith('"')) font = '"' + font;
+                    if (!font.endsWith('"')) font = font + '"';
+                    return font;
+                })
+                .join(',');
+        }
```
Comment (t3chguy) [inline src/settings/watchers/FontWatcher.ts:78]: Do we not need to clear it in the `else` case, the original code would write `""` in that case

File: src/settings/watchers/FontWatcher.ts [hunk 2]
```diff
@@ -64,6 +64,17 @@ export class FontWatcher implements IWatcher {
...
     private setSystemFont = ({ useSystemFont, font }) => {
-        document.body.style.fontFamily = useSystemFont ? font : "";
+        if (useSystemFont) {
+            // Make sure that fonts with spaces in their names get interpreted properly
+            document.body.style.fontFamily = font
+                .split(',')
+                .map(font => {
+                    font = font.trim();
+                    if (!font.startsWith('"')) font = '"' + font;
+                    if (!font.endsWith('"')) font = font + '"';
```
Comment (t3chguy) [inline src/settings/watchers/FontWatcher.ts:74]: Not sure these should be handled separately, maybe just a single case for if doesn't start&end with `"`

Comment (r00ster91) [reply src/settings/watchers/FontWatcher.ts]: I just did this for flexibility but yeah maybe this is not needed.

**Comment by t3chguy** [comment]
Do you think it'd be feasible to write a small test for this improvement?

**Comment by r00ster91** [comment]
@t3chguy Sadly I don't have any idea where or how to do that...
I suppose I would add an `x-test.ts` in `test/` but I don't know how to interact with the APIs to create that test.

**Comment by t3chguy** [comment]
So yeah you'd create a file `test/settings/watchers/FontWatcher-test.tsx` and test it via hitting `.start()` whilst mocking/stubbing the SettingsStore to manipulate the values for `useSystemFont` and `font` then `expect(...)`ing specific values to be set on `document.body.style.fontFamily`

You can probably copy `ThemeWatcher-test.tsx` from that same directory

**Comment by r00ster91** [comment]
I've added a basic template. How do I have to modify `SettingsStore` to set `document.body.style.fontFamily`?

**Comment by t3chguy** [comment]
@r00ster91 look at the other test in the same directory, it does these exact things

**Comment by r00ster91** [comment]
I wonder if something like this would do it.

File: test/settings/watchers/FontWatcher-test.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                expect(SettingsStore.getValue("systemFont")).toBe(`"Fira Sans Thin","Commodore 64"`);
+            });
+            SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, `"Commodore 64"`).then(() => {
+                expect(SettingsStore.getValue("systemFont")).toBe(`"Commodore 64"`);
+            });
+            SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, `  Fira Code  ,  "   Commodore 64   " `).then(() => {
+                expect(SettingsStore.getValue("systemFont")).toBe(`"Fira Code","Commodore 64"`);
+            });
+        });
+    });
```
[the remark this answers is not in the document]
Comment (r00ster91) [reply test/settings/watchers/FontWatcher-test.tsx]: I hope this resolved your points.

**Review by t3chguy — APPROVED** [review]
Awesome, thanks for your contribution!!

File: test/settings/watchers/FontWatcher-test.tsx [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+limitations under the License.
+*/
+
+import SettingsStore from '../../../src/settings/SettingsStore';
+import { SettingLevel } from '../../../src/settings/SettingLevel';
+
+describe('FontWatcher', function() {
+    beforeEach(() => SettingsStore.setValue("useSystemFont", null, SettingLevel.DEVICE, true));
+
+    it('encloses the fonts by double quotes and sets them as the system font', () => {
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:24]:
```suggestion
    it('encloses the fonts by double quotes and sets them as the system font', async () => {
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2022 r00ster91
+Copyright 2022 The Matrix.org Foundation C.I.C.
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:3]:
```suggestion
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import { SettingLevel } from '../../../src/settings/SettingLevel';
+
+describe('FontWatcher', function() {
+    beforeEach(() => SettingsStore.setValue("useSystemFont", null, SettingLevel.DEVICE, true));
+
+    it('encloses the fonts by double quotes and sets them as the system font', () => {
+        await SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, "Fira Sans Thin, Commodore 64");
+        expect(SettingsStore.getValue("systemFont")).toBe(`"Fira Sans Thin","Commodore 64"`);
+    });
+    it('does not add double quotes if already present and sets the font as the system font', () => {
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:28]:
```suggestion
    it('does not add double quotes if already present and sets the font as the system font', async () => {
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        await SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, `"Commodore 64"`);
+        expect(SettingsStore.getValue("systemFont")).toBe(`"Commodore 64"`);
+    });
+    it('trims whitespace, encloses the fonts by double quotes, and sets them as the system font', () => {
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:32]:
```suggestion
    it('trims whitespace, encloses the fonts by double quotes, and sets them as the system font', async () => {
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 8]
```diff
@@ -0,0 +1,36 @@
+/*
+Copyright 2022 r00ster91
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:2]:
```suggestion
Copyright 2022 r00ster91 <r00ster91@proton.me>
```

**Comment by r00ster91** [comment]
Right, I just wanted to add those `async`s as well.

File: test/settings/watchers/FontWatcher-test.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+import SettingsStore from '../../../src/settings/SettingsStore';
+import { SettingLevel } from '../../../src/settings/SettingLevel';
+
+describe('FontWatcher', function() {
+    beforeEach(() => SettingsStore.setValue("useSystemFont", null, SettingLevel.DEVICE, true));
+
+    it('encloses the fonts by double quotes and sets them as the system font', async () => {
+        await SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, "Fira Sans Thin, Commodore 64");
+        expect(SettingsStore.getValue("systemFont")).toBe(`"Fira Sans Thin","Commodore 64"`);
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:25]:
```suggestion
        expect(document.body.style.fontFamily).toBe(`"Fira Sans Thin","Commodore 64"`);
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    });
+    it('does not add double quotes if already present and sets the font as the system font', async () => {
+        await SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, `"Commodore 64"`);
+        expect(SettingsStore.getValue("systemFont")).toBe(`"Commodore 64"`);
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:29]:
```suggestion
        expect(document.body.style.fontFamily).toBe(`"Commodore 64"`);
```

File: test/settings/watchers/FontWatcher-test.tsx [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    });
+    it('trims whitespace, encloses the fonts by double quotes, and sets them as the system font', async () => {
+        const font = `  Fira Code  ,  "   Commodore 64   " `;
+        await SettingsStore.setValue("systemFont", null, SettingLevel.DEVICE, font);
+        expect(SettingsStore.getValue("systemFont")).toBe(`"Fira Code","Commodore 64"`);
```
Comment (t3chguy) [inline test/settings/watchers/FontWatcher-test.tsx:34]:
```suggestion
        expect(document.body.style.fontFamily).toBe(`"Fira Code","Commodore 64"`);
```

**Comment by t3chguy** [comment]
Async stuff is always fun, gimme a sec

**Comment by t3chguy** [comment]
Hey @r00ster91 unfortunately one of your tests has a legitimate failure

image

Which makes sense given both sides of the `"` have spaces on them, and trim will only take out the outer ones.

image

I'm going to update it to get rid of the spurious spaces inside the quotes, if a user put spaces in quotes they probably intended them..

---

Please review my changes and let me know if you're happy for them to land alongside your changes :)

**Review by r00ster91 — COMMENTED** [review]
Uh-huh, it seems you did some organizational changes too.
I'm surprised to see this PR turn out much more complicated than I initially thought.
It looks fine overall! You might as well add your copyright too.

**Comment by t3chguy** [comment]
> It looks fine overall! You might as well add your copyright too.

Haha, I'm about to add some tests into the file as part of another PR https://github.com/matrix-org/matrix-react-sdk/pull/8670 so will do so there

> Uh-huh, it seems you did some organizational changes too.

Mostly just to reuse testing utilities around the dispatcher

**Comment by t3chguy** [comment]
Thanks for writing the actual code to make custom fonts less confusing!!

**Pull request merged**